runtime/jcl/common/shared.c

```diff
@@ -561,7 +561,7 @@ getCpeTypeForProtocol(char* protocol, jsize protocolLen, const char* pathChars,
 	}
 	if (strncmp(protocol, "file", 5)==0) {
 		char* endsWith = (char*)(pathChars + (pathLen-4));
-		if ((strncmp(endsWith, ".jar", 4)==0) || (strncmp(endsWith, ".zip", 4)==0)) {
+		if ((strncmp(endsWith, ".jar", 4)==0) || (strncmp(endsWith, ".zip", 4)==0) || strstr(pathChars,".jar!/") || strstr(pathChars,".zip!/")) {
```

Comment:
On second look, can this be made more robust by just checking for `!/`

Reply:
Checking "!/" should also work here. 
We used to support only jar and zip, checking "!/" only means now we support other file format that contains "!/" even if it is not jar or zip. I am not sure if this will cause any problems.

Reply:
I guess it is fine, as long as Java can load classes from this file.

Reply:
It also means the code doesn't need to worry about multiple levels of nested jars, as `strstr(!/)` will always find the first level.